File: Tracks/csharp/beer-song/BeerSong.cs
using System.Linq;

public static class BeerSong
{
    private const string ITEM = "bottle";
    private const string CONTENT = "beer";
    private const string LOCATION = "wall";
    private static string ItemsOfContent(int left, bool onTheWall = false) => $"{Items(left)} of {CONTENT}{(onTheWall ? $" on the {LOCATION}" : "")}";

    private static string Items(int left) => left switch
    {
        1 => $"{left} {ITEM}",
        0 => $"No more {ITEM}s",
        _ => $"{left} {ITEM}s",
    };

    private static string Single(int left) => left switch
    {
        1 => "it",
        _ => "one"
    };

    private static string WhatDoWeDoWithThese(int left) => left switch
    {
        0 => $"Go to the store and buy some more, {ItemsOfContent(99, onTheWall: true).ToLower()}.",
        _ => $"Take {Single(left)} down and pass it around, {ItemsOfContent(left - 1, onTheWall: true).ToLower()}."
    };

    public static string Recite(int startBottles, int takeDown) => string.Join("\n\n", Enumerable.Range(0, takeDown).Select((index) =>
        {
            int left = startBottles - index;
            return $"{ItemsOfContent(left, onTheWall: true)}, {ItemsOfContent(left).ToLower()}.\n{WhatDoWeDoWithThese(left)}";
        }));
}

File: Tracks/typescript/beer-song/beer-song.ts
const howMany = (left: number): string => (left > 1 ? "one" : "it");
const howManyLeft = (left: number, upper = false): string =>
  (left === 0
    ? `${upper ? "N" : "n"}o more bottles`
    : `${left} bottle${left > 1 ? "s" : ""}`) + " of beer";

export default class Beer {
  static verse(left: number): string {
    const firstLine = `${howManyLeft(left, true)} on the wall, ${howManyLeft(
      left
    )}.`;
    const secondPart = `${howManyLeft(left > 0 ? left - 1 : 99)} on the wall`;
    const secondLine =
      left > 0
        ? `Take ${howMany(left)} down and pass it around, ${secondPart}.`
        : `Go to the store and buy some more, ${secondPart}.`;

    return firstLine.concat("\n", secondLine, "\n");
  }
  static sing = (from: number = 99, to: number = 0): string =>
    [...Array(from - to + 1)]
      .map((_, index) => Beer.verse(from - index))
      .join("\n");
}
